tracing-subscriber/src/fmt/fmt_layer.rs

```diff
@@ -458,6 +459,81 @@ where
         }
     }
 
+    fn on_enter(&self, id: &Id, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+        if let Some(timings) = extensions.get_mut::<Timings>() {
+            timings.entered = Some(Instant::now());
+        } else {
+            extensions.insert(Timings::new());
+        }
+    }
+
+    fn on_exit(&self, id: &Id, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+        if let Some(timings) = extensions.get_mut::<Timings>() {
+            if let Some(started) = timings.entered.take() {
+                timings.accumulated += started.elapsed().as_nanos() as u64;
+            }
+        }
+    }
+
+    fn on_close(&self, id: Id, ctx: Context<'_, S>) {
+        let span = ctx.span(&id).expect("Span not found, this is a bug");
+        let extensions = span.extensions();
+        if let Some(timing) = extensions.get::<Timings>() {
+            let t = field::display(timing.get_accumulated());
+            let meta = span.metadata();
+            let name = field::display(meta.name());
+            let cs = meta.callsite();
+            let meta = Metadata::new(
+                meta.name(),
+                meta.target(),
+                meta.level().clone(),
+                meta.file(),
+                meta.line(),
+                meta.module_path(),
+                field::FieldSet::new(&[], cs.clone()),
+                Kind::EVENT,
+            );
+            let fs = field::FieldSet::new(&["message", "timing"], cs);
+            let mut it = fs.iter();
+            let v = [
+                (&it.next().unwrap(), Some(&name as &dyn field::Value)),
+                (&it.next().unwrap(), Some(&t as &dyn field::Value)),
+            ];
+            let vs = fs.value_set(&v);
+            let event =
+                Event::new_child_of(id, unsafe { &*(&meta as *const Metadata<'static>) }, &vs);
```

Comment:
The unsafe cast here, however, _is_ UB. Even though the formatter's `format_event` function is _probably_ not doing anything that relies on the event's metadata being `'static`, we can't really guarantee that at all, since it receives metadata that has the `'static` lifetime. If the `format_event` impl _did_ do something like putting it in some other data structure that outlives this stack frame, and tried to access it later, it would be a dangling pointer. We should try to find a way to not do this.

Reply:
The reason why I did this is that I didn’t want the span’s fields duplicated in the event, but that didn’t work as intended anyway: the span is already closed when the event is generated, so the proper solution might be exactly to reuse all the meta-data. But I couldn’t get the value_set to come out, yet.

Reply:
I think you _should_ be able to just do exacly what you're doing now with the `ValueSet`. I think it's okay for the `FieldSet` on the metadata to declare different fields --- note that this code has that, as well, since the statically declared `FieldSet` is empty.

The metadata's `FieldSet` is mainly used for two things:
 - filtering, which (again), doesn't really matter here --- we probably want to synthesize these events for _any_ span that the filter has already chosen to enable, even if it wouldn't enable the event itself, so it's fine to "sneak past" filtering
 - caching `Field` values in the subscriber so that it can compare them without falling back to string matching. that's just an optimization, and if a subscriber is using it, it should degrade gracefully in this case --- the worst-case is that the subscriber may do a string comparison for these events that it wouldn't normally.